guava/src/com/google/common/base/Optional.java

```diff
@@ -282,7 +284,7 @@
    * @throws NullPointerException if the function returns {@code null}
    * @since 12.0
    */
-  public abstract <V> Optional<V> transform(Function<? super T, V> function);
+  public abstract <V extends @NonNull Object> Optional<V> transform(Function<? super T, V> function);
```

Comment:
I think that every occurrence of `extends @NonNull Object` in the diffs is unnecessary.  I suggest removing them.

Reply:
Explicit and Implicit bounds for generics are treated differently by the Checker Framework. Here is a link to documentation that explains the difference, https://checkerframework.org/manual/#defaults-classfile.

By writing `V extends @NonNull Object` we are restricting the instantiation to `@NonNull` types which would have otherwise defaulted to type `@Nullable` as per the Checker Framework's defaulting rules. For more information on restricting the instantiation of generic class please the refer to the link: https://checkerframework.org/manual/#generics

If the usage `extends @NonNull Object` is logically incorrect then please let me know, so that I can remove these annotations in that case. Further, I think every occurrence of `extends @Nullable Object` is redundant and can be removed.

Thanks!

Reply:
> Explicit and Implicit bounds for generics are treated differently by the Checker Framework. Here is a link to documentation that explains the difference, https://checkerframework.org/manual/#faq-semantics-section.

You gave a link to a section about unannotated libraries.  This is not relevant to your work.

To support your claim, it would be helpful for you to give a concrete example of code in which explicit and implicit bounds are treated differently by the Checker Framework.

In particular, you coud write code like this:

```
   <V extends @NonNull Object> ReturnType methodName(args...) { ... }
   <V extends @Nullable Object> ReturnType methodName(args...) { ... }
   <V extends Object> ReturnType methodName(args...) { ... }
```

where everything after the type argument declaration (that is, everything starting with `ReturnType`) is identical, but the Checker Framework treats all three methods differently when type-checking clients.

Reply:
Please consider the following given example
```java
import org.checkerframework.checker.nullness.qual.NonNull;
import org.checkerframework.checker.nullness.qual.Nullable;

public class Test {
  public static <V extends @NonNull Object> String dummyA(V argA) {
    return argA.toString();
  }
  
  public static <V extends @Nullable Object> String dummyB(V argB) {
    return argB.toString();
  }

  public static <V extends Object> String dummyC(V argC) {
    return argC.toString();
  }

  public static <V> String dummyD(V argD) {
    return argD.toString();
  }

  public static <V extends @Nullable Object> V dummyE(V argE) {
    return null;
  }

  public static <@Nullable V extends @Nullable Object> V dummyF(V argF) {
    return null;
  }

  public static void main(String[] args) {
    @Nullable String nullableStr = null;
    String nonnullStr = "Dummy";
    
    dummyA(nullableStr);
    dummyB(nullableStr);
    dummyC(nullableStr);
    dummyD(nullableStr);

    dummyA(nonnullStr);
    dummyB(nonnullStr);
    dummyC(nonnullStr);
    dummyD(nonnullStr);
  }
}
```
Following depicts the compile-time errors thrown by nullness type processor
```
Test.java:10: error: [dereference.of.nullable] dereference of possibly-null reference argB
    return argB.toString();
           ^
Test.java:18: error: [dereference.of.nullable] dereference of possibly-null reference argD
    return argD.toString();
           ^
Test.java:22: error: [return.type.incompatible] incompatible types in return.
    return null;
           ^
  found   : @FBCBottom @Nullable  null
  required: V[ extends @Initialized @Nullable Object super @Initialized @NonNull Void]
Test.java:33: error: [type.argument.type.incompatible] incompatible types in type argument.
    dummyA(nullableStr);
          ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull Object
Test.java:35: error: [type.argument.type.incompatible] incompatible types in type argument.
    dummyC(nullableStr);
          ^
  found   : @Initialized @Nullable String
  required: @Initialized @NonNull Object
5 errors
```
Conclusion
* `@NonNull` is treated as the implicit lower bound. Evident from the error message for dummyE which returns null as a value.
* Both `V extends @NonNull Object` and `V extends Object` are equivalent and are treated as `V [extends @NonNull Object super @NonNull Void]` , these both restrict the type V to `@NonNull` types only. Thus passing a `@Nullable` type reference causes checker to throw an error, as evident from the error message for method dummyA and dummyC.
* Both `V` and `V extends @Nullable Object` are identical and are treated as `V [extends @Nullable Object super @NonNull Void]`

This is in accordance with the defaulting rules documented in the manual, see, https://checkerframework.org/manual/#climb-to-top, which says the following about implicit upper and lower bounds
>Implicit upper bounds are defaulted to top to allow them to be instantiated in any way. If a user declared class C\<T\> { ... }, then we assume that the user intended to allow any instantiation of the class, and the declaration is interpreted as class C\<T extends `@Nullable` Object\> { ... } rather than as class C\<T extends `@NonNull` Object\> { ... }.

>Implicit lower bounds are defaulted to the bottom type, again to allow maximal instantiation. Note that Java does not allow a programmer to express both the upper and lower bounds of a type, but the Checker Framework allows the programmer to specify either or both

Thanks!